java/client/src/org/openqa/selenium/remote/RemoteWebDriver.java

```diff
@@ -155,7 +155,7 @@ public static RemoteWebDriverBuilder builder() {
   }
 
   private void init(Capabilities capabilities) {
-    capabilities = capabilities == null ? new ImmutableCapabilities() : capabilities;
+    this.capabilities = capabilities == null ? new ImmutableCapabilities() : capabilities;
```

Comment:
Assigning capabilities to the field here is not the best idea. Semantically this field contains the capabilities returned by the browser after session start. So here we sould better amend capabilities and return them from `init` method to pass later to `startSession` method (that will assign the capabilities returned by the browser to the field).

Reply:
And instead of assigning capabilities to the field prior to a session start we can modify `getCapabilities` method to return an empty `ImmutableCapabilities` object if the field is equal to `null` to avoid `NullPointerException` here.

Reply:
Don't you think it will be overhead of null check each time `getCapabilities` called? Ideally `init` accepts desired capabilities and capabilities get overridden with [actual capabilities](https://github.com/SeleniumHQ/selenium/blob/30f9ffab92839cdd454b65404b50ee94f887e99e/java/client/src/org/openqa/selenium/remote/RemoteWebDriver.java#L219) in [startSession](https://github.com/SeleniumHQ/selenium/blob/30f9ffab92839cdd454b65404b50ee94f887e99e/java/client/src/org/openqa/selenium/remote/RemoteWebDriver.java#L255). So assigning empty `ImmutableCapabilities` capabilities in `init` should be fine. If we don't want to set `ImmutableCapabilities` in `init`, better is to add setter method instead of null check on each get call.

Reply:
People will hardly call `getCapabilities` often in a non-initialized driver, and this field will be set in `startSession` to the capabilities returned by the browser after initialization

Reply:
Execute script command call will call capabilities each time before executing execute script command.

Reply:
I want to reiterate that this field will be initialized after a call to `startSession` method, it will contain the capabilities returned by the browser:
https://github.com/SeleniumHQ/selenium/blob/30f9ffab92839cdd454b65404b50ee94f887e99e/java/client/src/org/openqa/selenium/remote/RemoteWebDriver.java#L255